api/osrd_infra/migrations/__init__.py

```diff
@@ -35,9 +50,209 @@ def run_sql_add_foreign_key(model_name: str, field_name: str, link_model: str, n
+                {query_continuation};
+            RETURN NEW;
+        END;
+        $$;
+        CREATE TRIGGER {trigger}
+        AFTER INSERT
+        ON "{source_table}"
+        FOR EACH ROW
+        EXECUTE FUNCTION {proc}();
+    """.format(
+            source="t"
+        ),
+        f"""DROP TRIGGER IF EXISTS {trigger} ON "{source_table}";
+        DROP FUNCTION IF EXISTS {proc};
+    """,
+    )
+
+
+def update_trigger(search_table: str, source_table: str, columns: dict[str, Sql], query_continuation: Sql) -> Sql:
+    set_columns = ", ".join(f'"{col}" = ({sql})' for col, sql in columns.items())
+    trigger = f'"{search_table}__upd_trig"'
+    proc = f'"{search_table}__upd_trig_fun"'
+    return (
+        f"""
+        CREATE OR REPLACE FUNCTION {proc}()
+            RETURNS TRIGGER
+            LANGUAGE plpgsql
+        AS $$
+        BEGIN
+            UPDATE "{search_table}"
+                SET {set_columns}
+                FROM (SELECT NEW.*) AS t
+                WHERE t.id = "{search_table}".id
+                {query_continuation};
+            RETURN NEW;
+        END;
+        $$;
+        CREATE TRIGGER {trigger}
+        AFTER UPDATE
+        ON "{source_table}"
+        FOR EACH ROW
+        EXECUTE FUNCTION {proc}();
+    """.format(
+            source="t"
+        ),
+        f"""DROP TRIGGER IF EXISTS {trigger} ON "{source_table}";
+        DROP FUNCTION IF EXISTS {proc};
+    """,
+    )
+
+
+def run_sql_create_infra_search_table(
+    name: str,
+    source_table: Sql,
+    search_columns: dict[str, Sql],
+    *,
+    extra_columns: Optional[dict[str, tuple[Sql, Sql]]] = None,
+    table_type: SearchTableType = SearchTableType.TABLE,
+    index: SearchIndex = SearchIndex.GIN,
+    phony_model_name: Optional[str] = None,
+    source_table_pk: str = "id",
+    triggers: bool = False,
+    create_table_continuation: Sql = "",
+    query_continuation: Sql = "",
+) -> migrations.RunSQL:
+    """Creates a migration that setup a table optimized for searching.
+
+    Example:
+
+    ```
+    run_sql_create_infra_search_table(
+        name="my_search",
+        source_table="clients",
+        search_columns={
+            "fn": "first_name",
+            "ln": "last_name",
+            "full_name": "first_name || ' ' || last_name",
+            "age": "{source}.age",
+        },
+        extra_columns={"non_indexed_column", ("12", "INT")},
+    )
+    ```
+
+    Will create a view named `my_search` getting its data from the schema `clients` with:
+
+    - An id column which references `source_table_pk`
+    - 4 columns indexed for optimized text searching: fn, ln, full_name and age.
+      Their value before optimization correspond the evaluated SQL expression in
+      the `column` dict.  That SQL expression's type must be TEXT.  When the column
+      name and its expression can be ambiguous (eg. age), the str.format pattern `{source}`
+      is supported and will be filled accordingly as to avoid ambuiguities.
```

Comment:
```suggestion
      is supported and will be filled accordingly as to avoid ambiguities.
```
This extension is quite good to avoid misspelling : https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker